nextest-runner/src/test_filter/expression.rs

```diff
@@ -0,0 +1,128 @@
+#[derive(Debug, Clone)]
+pub enum NameMatcher {
+    /// Exact value
+    Equal(String),
+    /// Simple contains test
+    Contains(String),
+    /// Test against a regex
+    Regex(regex::Regex),
+}
+
+#[cfg(test)]
+impl PartialEq for NameMatcher {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Contains(s1), Self::Contains(s2)) => s1 == s2,
+            (Self::Equal(s1), Self::Equal(s2)) => s1 == s2,
+            (Self::Regex(r1), Self::Regex(r2)) => r1.as_str() == r2.as_str(),
+            _ => false,
+        }
+    }
+}
+
+#[cfg(test)]
+impl Eq for NameMatcher {}
+
+/// Define a set of tests
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[derive(Debug, Clone)]
+pub enum SetDef {
+    /// All tests in a package
+    Package(NameMatcher),
+    /// All tests in a package dependencies
+    Deps(NameMatcher),
+    /// All tests in a package reverse dependencies
+    Rdeps(NameMatcher),
+    /// All tests matching a name
+    Test(NameMatcher),
+    /// All tests
+    All,
+    /// No tests
+    None,
+    // Possible addition: Binary(NameMatcher)
+}
+
+/// Filtering expression
+///
+/// Used to filter tests to run.
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[derive(Debug, Clone)]
+pub enum Expr {
+    /// Accepts every tests not in the given expression
+    Not(Box<Expr>),
+    /// Accepts every tests in either given expression
+    Union(Box<Expr>, Box<Expr>),
+    /// Accepts every tests in both given expression
+    Intersection(Box<Expr>, Box<Expr>),
+    /// Accepts every tests in a set
+    Set(SetDef),
+}
+
+impl NameMatcher {
+    pub fn is_match(&self, input: &str) -> bool {
+        match self {
+            Self::Equal(text) => text == input,
+            Self::Contains(text) => input.contains(text),
+            Self::Regex(reg) => reg.is_match(input),
+        }
+    }
+}
+
+impl SetDef {
+    fn includes(&self, artifact: &RustTestArtifact<'_>, name: &str) -> bool {
+        match self {
+            Self::All => true,
+            Self::None => false,
+            Self::Test(matcher) => matcher.is_match(name),
+            Self::Package(matcher) => matcher.is_match(artifact.package.name()),
+            Self::Deps(matcher) => Self::matches_dependency_of(artifact, matcher),
+            Self::Rdeps(_matcher) => unimplemented!("rdeps not implemented"),
+        }
+    }
+
+    /// Returns true if `artifact` is a dependency of at least
+    /// one workspace package matching `matcher`.
+    fn matches_dependency_of(artifact: &RustTestArtifact<'_>, matcher: &NameMatcher) -> bool {
+        let graph = artifact.package.graph();
+        let mut packages = graph
+            .packages()
+            .filter(|p| p.in_workspace())
+            .filter(|p| matcher.is_match(p.name()));
+
+        let id = artifact.package.id();
+        let mut cache = graph.new_depends_cache();
+
+        packages.any(|p| p.id() != id && cache.depends_on(p.id(), id).unwrap_or(false))
+    }
+}
```

Comment:
This seems like it'll blow up the time complexity of the operation quite significantly:
* the number of artifacts is `O(number of workspace packages)`
* the complexity of graph traversal is `O(number of workspace packages + number of edges)`
* filtering by package is itself `O(number of workspace packages)`.

So it's worse than cubic in the number of workspace packages. For workspace-wide operations it's generally good to avoid anything past `O(n log n)` where `n` is the number of workspace packages.

Let's skip `deps` and `rdeps` operations for now and figure out how to implement them efficiently in a separate PR -- I think there's a bunch of ways we can rewrite this to make it more efficient.

Reply:
You're right, I didn't though about that.

And we (I) need to remember to remove `--no-deps` form our call to `cargo metadata`...

Reply:
We might "compile" the parsed expression using the packages graph to replace `deps()` and `rdeps()` by a set of accepted package id. This would reduce the cost inside `includes(artifact, test_name)`

Reply:
Rework with the compile approach. Let me known If you think this is better.